packages/frontend/@n8n/design-system/src/components/TableHeaderControlsButton/TableHeaderControlsButton.vue

```diff
@@ -0,0 +1,301 @@
+<script setup lang="ts" generic="ColumnType extends ColumnHeader">
+import { computed, ref } from 'vue';
+
+import { useI18n } from '@n8n/design-system/composables/useI18n';
+import type { ButtonSize, IconSize } from '@n8n/design-system/types';
+
+import N8nButton from '../N8nButton';
+import N8nIcon from '../N8nIcon';
+import N8nPopoverReka from '../N8nPopoverReka/N8nPopoverReka.vue';
+
+export type ColumnHeader =
+	| {
+			key: string;
+			label: string;
+			visible: boolean;
+			disabled: false;
+	  }
+	| { key: string; disabled: true };
+
+interface Props {
+	columns: ColumnType[];
+	buttonSize?: ButtonSize;
+	iconSize?: IconSize;
+}
+
+const props = defineProps<Props>();
+
+const visibleColumns = computed(() =>
+	props.columns.filter(
+		(column): column is ColumnType & { disabled: false } => !column.disabled && column.visible,
+	),
+);
+const hiddenColumns = computed(() =>
+	props.columns.filter(
+		(column): column is ColumnType & { disabled: false } => !column.disabled && !column.visible,
+	),
+);
+
+const { t } = useI18n();
+
+const draggedItem = ref<string | null>(null);
+const dragOverItem = ref<string | null>(null);
+
+const emit = defineEmits<{
+	'update:columnVisibility': [key: string, visibility: boolean];
+	'update:columnOrder': [newOrder: string[]];
+}>();
+
+const resetDragState = () => {
+	draggedItem.value = null;
+	dragOverItem.value = null;
+};
+
+const handleDragStart = (event: DragEvent, columnKey: string) => {
+	if (!event.dataTransfer) return;
+	draggedItem.value = columnKey;
+	event.dataTransfer.effectAllowed = 'move';
+	event.dataTransfer.setData('text/plain', columnKey);
```

Comment:
Do we need `columnKey` both in `draggedItem` and `dataTransfer`?

Reply:
Ya I guess we are not using the data in the event here. Although some browsers (particularly older versions of Firefox and Safari) will not allow drag operations to proceed unless `data` is set in the dataTransfer object.